File: leetcode/0015_3sum/0015.go
package threesum

import "sort"

// ThreeSum finds all unique triplets in the array which give the sum of zero
func ThreeSum(nums []int) [][]int {
	var res [][]int
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		if i == 0 || nums[i] != nums[i-1] {
			// j -- lower bound for bi-directional sweep
			// k -- upper bound
			sum, j, k := -nums[i], i+1, len(nums)-1
			for j < k {
				if nums[j]+nums[k] == sum {
					res = append(res, []int{nums[i], nums[j], nums[k]})
					for j < k && nums[j] == nums[j+1] {
						j++
					}
					for j < k && nums[k] == nums[k-1] {
						k--
					}
					j++
					k--
				} else if nums[j]+nums[k] < sum {
					j++
				} else {
					k--
				}
			}
		}
	}
	return res
}

File: leetcode/0015_3sum/0015.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(begin(nums), end(nums));
        for (int i = 0; i < int(nums.size()) - 2; ++i) {
            if (nums[i] > 0) return res; // pruning
            if (i == 0 || nums[i] != nums[i - 1]) {
                int sum = -nums[i], j = i + 1, k = nums.size() - 1;
                while (j < k) {
                    if (nums[j] + nums[k] == sum) {
                        res.push_back({nums[i], nums[j], nums[k]});
                        while (j < k && nums[j] == nums[j + 1]) ++j;
                        while (j < k && nums[k] == nums[k - 1]) --k;
                        ++j;
                        --k;
                    } else if (nums[j] + nums[k] < sum) ++j;
                    else --k;
                }
            }
        }
        return res;
    }
};
